### ISSUE TITLE
WFSCollider won't let me change drivers and Hardware buffersize problem

### ISSUE THREAD

**Issue by Polkjia** [GameOfLife/WFSCollider-Class-Library#4]
Hello there,

I managed to install WFSC and making it work, but I have two problems:

1) with MME drivers I am obliged to set to 1024 the Hardware Buffer size, otherwise it smears and slows audio even with 30 second file, no heavy processes. PC is definitely able to handle everything. 
2) If I try to change used driver, there is only 'default' and no list. Even if I set my drivers directly from SC prior to start, WFSC obliges me to use MME. I can then reboot SC with the right drivers, and it is even working, but SC server's GUI numbers (bottom right) seems to not work at all. 

Regards

**Comment by woutersnoei** [comment]
Hi Polkjia, I don't have access to a windows machine to test things, so would you be ok to try some things for me? If yes, first up; could you try starting up WFSCollider (WFSLib.startup), let the server startup, then quit it (using the green 'power' button), run the following code:
s.options.inDevice = "--name of your input device--";
s.options.outDevice = "--name of your output device--";
and then start the server again?
If I'm correct this should be a workaround for the time being. Main issue, also for macOS, is the fact that in and out devices can have different names. This used to be rare on macos, but is now normal. Also for windows SC only recently added the option to request device names from the lang. This would create an error before, so I disabled it for windows (hence only "default" in the menu). I can put it back in, but need to rework it to make difference between input and output device names possible. If the above works for you I know what to do to make it work with the preference panel as well. Cheers & thanks, Wouter

**Comment by Polkjia** [comment]
Dear Wouter,

It works. 
![image](https://user-images.githubusercontent.com/77069555/147830922-c1c8e653-238d-4c45-a2af-ca23bd99c5b5.png)


The only thing is that the server in SC doesn't appear active, even if everything from WFS works. I am not sure if this is correct.

Plus, whenever you feel like to test on a windows machine with a windows user, please feel free to contact me. I will gladly help you. 
Cheers and happy last day of the year!

**Comment by woutersnoei** [comment]
Hi Polkjia,
and a very happy new year to you!
I just updated the WFSCollider Class Library and Unit Lib to (hopefully) enable correct workings of the device settings on all platforms. Let me know if it works for you! 
Cheers, Wouter

**Comment by Polkjia** [comment]
Hello Wouter
It doesn't work, I see this error

> compiling class library...
	Found 853 primitives.
	Compiling directory 'C:\Program Files\SuperCollider-3.12.1\SCClassLibrary'
	Compiling directory 'C:\ProgramData\SuperCollider\Extensions'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\Extensions'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\VectorSpace'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\XML'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\wslib'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\Unit-Lib'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\PopUpTreeMenu'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\ArdourOSC'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\adclib'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\JITLibExtensions'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\Utopia'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\HyperDisCo'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\AllGui'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\JACK'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\MathLib'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\NetLib'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\SafetyNet'
	Compiling directory 'C:\Users\Polka\AppData\Local\SuperCollider\downloaded-quarks\Vowel'
	numentries = 2149265 / 54353468 = 0.04
	11414 method selectors, 4762 classes
	method table size 45501168 bytes, big table size 434827744
	Number of Symbols 26074
	Byte Code Size 1121981
	compiled 1165 files in 1.96 seconds

> Info: 29 methods are currently overwritten by extensions. To see which, execute:
MethodOverride.printAll
compile done
localhost : setting clientID to 0.
internal : setting clientID to 0.
Safety('internal') enabled.
Safety('localhost') enabled.
Execution warning: Class 'UAudioDeviceSpec' not found
ERROR: Message 'new' not understood.
RECEIVER:
nil
ARGS:
Instance of String {    (00000292513D9C40, gc=01, fmt=07, flg=11, set=02)
  indexed slots [10]
      0 : J
      1 : a
      2 : c
      3 : k
      4 : R
      5 : o
      6 : u
      7 : t
      8 : e
      9 : r
}
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand
		arg this = nil
		arg selector = 'new'
		arg args = [*1]
	Meta_WFSServerOptionsGUI:initClass
		arg this = <instance of Meta_WFSServerOptionsGUI>
	Meta_Class:initClassTree
		arg this = <instance of Meta_Class>
		arg aClass = <instance of Meta_WFSServerOptionsGUI>
		var implementsInitClass = nil
	ArrayedCollection:do
		arg this = [*3]
		arg function = <instance of Function>
		var i = 0
	Meta_Class:initClassTree
		arg this = <instance of Meta_Class>
		arg aClass = <instance of Meta_AbstractWFSOptionsGUI>
		var implementsInitClass = nil
	ArrayedCollection:do
		arg this = [*568]
		arg function = <instance of Function>
		var i = 376
	Meta_Class:initClassTree
		arg this = <instance of Meta_Class>
		arg aClass = <instance of Meta_Object>
		var implementsInitClass = nil
	Process:startup
		arg this = <instance of Main>
		var time = 1.9945236
	Main:startup
		arg this = <instance of Main>
		var didWarnOverwrite = false
^^ ERROR: Message 'new' not understood.
RECEIVER: nil


> SCDoc: Indexing help-files...
WARNING: SCDoc: ArdourOSC has no categories!
SCDoc: Indexed 2640 documents in 2.0 seconds


When deleting the folder, it behaves normally.

**Comment by woutersnoei** [comment]
Ah, sorry forgot to mention; you need to update Unit-Lib as well, then it should work.

**Comment by Polkjia** [comment]
I guess I should do it manually since I cannot access Quarks with this errors?

**Comment by woutersnoei** [comment]
Or manually remove wfs lib from the installed quarks folder and then recompile to prevent the errors? (And then later install the wfs lib quark again)

**Comment by Polkjia** [comment]
I am not using WFSC quark because it doesn't download/show any class on my machine (see other issue), I am using Unit-Lib quarks. 
However, I uninstalled Unit-Lib quark and then proceeded to manually install both WFS class lib and Unit-Lib. It works. 

I also tested with Jack and it works, also with very CPU consuming settings (Hardware Buffer Size 64 and block size 64): it works without even flinching. I can confirm MME drivers do not work well with WFS, and requires higher HardBuffSize. 

Ofc thank you a lot for your help!
